inc/Model.h

```diff
@@ -1,85 +1,224 @@
-/**  Model - Header
-*   @file       Model.h
-*   @ingroup    GSAP-Support
-*
-*   @brief      Model class
-*
-*   @author     Matthew Daigle
-*   @version    0.1.0
-*
-*   @pre        N/A
-*
-*      Contact: Matthew Daigle (matthew.j.daigle@nasa.gov)
-*      Created: March 5, 2016
-*
-*   @copyright Copyright (c) 2018 United States Government as represented by
-*     the Administrator of the National Aeronautics and Space Administration.
-*     All Rights Reserved.
-*/
-
+// Copyright (c) 2017-2018 United States Government as represented by the
+// Administrator of the National Aeronautics and Space Administration.
+// All Rights Reserved.
 #ifndef PCOE_MODEL_H
 #define PCOE_MODEL_H
 
-#include <vector>
 #include <string>
+#include <vector>
+
+#include "DynamicArray.h"
```

Comment:
Remind me, what's the benefit here in having a dynamic array class instead of defining InputVector to be vector<double> (like noisetype)?

Reply:
Making each of the input types distinct by inheriting from a base type prevents accidental reordering of arguments that simple `using` statements don't.

```c++
class Model {
    using state_type = std::vector<double>;
    using input_type = std::vector<double>;

    void initialize(state_type x, input_type u);
};

int main() {
    Model m;
    state_type x;
    input_type u;
    m.initialize(u, x); // BAD: Not a compiler error!
}
```

versus

```c++
class StateVector: DynamicArray {};

class InputVector: DynamicArray {};

class Model {
    using state_type = StateVector;
    using input_type = InputVector;

    void initialize(state_type x, input_type u);
};

int main() {
    Model m;
    state_type x;
    input_type u;
    m.initialize(u, x); // Good: Compiler error
}
```

An alternative solution would be to create a struct with named vectors. I think the custom-type solution is better because it requires a bit of extra effort up front, but makes life easier for users.

Of course, `DynamicArray` is necessary because it is *wrong*™ to inherit from STL containers 😛

Reply:
Gotcha. That makes a lot of sense. Thanks for explaining it